File: SCTW-DS/scripts/Utils/progress_bar.py
import sys
import threading


class ProgressBar:
    def __init__(
        self,
        total: int,
        program_name: str,
        bar_length=20,
        done_message="Finished!",
        update_interval=1,
        cost=None,
    ):
        self.total = total
        self.bar_length = bar_length
        self.done_message = done_message
        self.update_interval = update_interval
        self.current = 0
        self.lock = threading.Lock()

        print(f"\n\x1b[32m----Starting {program_name}----\x1b[0m\n")

        self.update(0, cost)

    def update(self, current: int, cost):
        with self.lock:
            self.current = current
            fraction = self.current / self.total

            arrow = int(fraction * self.bar_length) * "-" + ">"
            padding = (self.bar_length - len(arrow)) * " "

            if self.current == self.total:
                ending = f"\n\n{self.done_message}\x1b[0m\n\n"
            else:
                ending = "\r"

            completed = "\x1b[32m" if self.current == self.total else "\x1b[0m"

            print(
                f"{completed}Progress: [{arrow}{padding}] {fraction*100:.2f}% Cost: {cost}",
                end=ending,
            )
            sys.stdout.flush()

    def increment(self, step=1, cost=None):
        self.update(self.current + step, cost)

File: SCTW_DS/src/NeuralNetwork/misc/ProgressBar.cpp
#include <iostream>
#include <string>

using namespace std;

class ProgressBar {
 private:
  int total;
  int current;
  int bar_length;
  string message;

 public:
  ProgressBar(int total, int bar_length = 20,
              const string &message = "Finished!")
      : total(total), current(0), bar_length(bar_length), message(message) {
    std::cout << "\n\x1b[32m----Starting Training----\x1b[0m\n" << std::endl;
    update(0, 0.0);
  }

  void update(int current, double cost) {
    double fraction = static_cast<double>(current) / total;

    string arrow;
    for (int i = 0; i < static_cast<int>(fraction * bar_length) - 1; ++i) {
      arrow += "-";
    }
    arrow += ">";

    string padding;
    for (int i = 0; i < bar_length - arrow.size(); ++i) {
      padding += " ";
    }

    string ending = current == total ? "\n\n" + message + "\n\n" : "\r";
    string completed = current == total ? "\x1b[32m" : "\x1b[0m";

    cout << completed << "Progress: [" << arrow << padding << "] "
         << fraction * 100 << "%" << " Cost: " << cost << ending;
  }

  void increment(int step = 1, double cost = 0.0) {
    current += step;
    update(current, cost);
  }
};
